[WIP] Intrumentation: Instrumentation for locks.

hackintoshrao opened the pull request:
- Enabled by setting MINIO_DEBUG=lock. 
- Info avaialbe via endpoint, currently active on `/minio/debug/locks`
- Here is the current format

```
{
    "totalLocks": 1,
    "totalBlockedLocks": 0,
    "totalAcquiredLocks": 1,
    "locksInfoPerObject": [{
        "bucket": "testbucket",
        "object": "1gb",
        "locksOnObject": 1,
        "locksAcquiredOnObject": 1,
        "locksBlockedOnObject": 0,
        "LockDetailsOnObject": [{
            "opsID": "MCPNVX9ULMV16Q5S",
            "lockOrigin": "[lock held] in main.xlObjects.GetObject[/home/minio/gopath/src/github.com/minio/minio/xl-v1-object.go:64]",
            "lockType": "RLock",
            "status": "Running",
            "statusSince": "403.022271ms"
        }]
    }]
}
```

harshavardhana commented on main.go:
```diff
@@ -63,6 +63,11 @@ func init() {
 	// Set global trace flag.
 	globalTrace = os.Getenv("MINIO_TRACE") == "1"
+
+	switch os.Getenv("MINIO_DEBUG") {
```
why not do this? 

```
debugEnv := os.Getenv("MINIO_DEBUG") 

globalLockDebug = (debugEnv == "lock")
```

hackintoshrao replied:
In future MINIO_DEBUG can have multiple debugging options, `lock`, `health` and so on... so placed it in switch case statment.

harshavardhana replied:
> In future MINIO_DEBUG can have multiple debugging options, lock, health and so on... so placed it in switch case statment.

In that case move it to a function. 'setGlobalsDebugFromEnv()'

harshavardhana commented on lock-instrument.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// return randomly generated string ID if lock debug is enabled,
+// else returns empty string
```
```
func getOpsID() (opsID string) {
    if globalLockDebug {
           opsID = string(generatedRequestID())
     } 
     return opsID
}
```

harshavardhana commented on lock-instrument.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// Read entire state of the locks in the system and return.
+// [WIP]
```
why "WIP" ?

hackintoshrao replied:
That's left over. Will remove.

harshavardhana commented on fs-v1-multipart.go:
```diff
@@ -58,9 +58,13 @@ func (fs fsObjects) listMultipartUploads(bucket, prefix, keyMarker, uploadIDMark
-		nsMutex.RLock(minioMetaBucket, pathJoin(mpartMetaPrefix, bucket, keyMarker))
+		// generates random string on setting MINIO_DEBUG=lock, else returns empty string.
+		// used for instrumentation on locks.
+		opsID := getOpsID()
```
Still thinking about how to avoid OpsID() here.. do not wish to change the function signature of nsMutex.Lock(), nsMutex.RLock()

hackintoshrao replied:
Yes, It could be more cleaner if it can be avoided.

harshavardhana commented on server-main.go:
```diff
@@ -61,6 +61,22 @@ ENVIRONMENT VARIABLES:
+		 Example: export  MINIO_CACHE_EXPIRY=10hours
+		          export  MINIO_CACHE_SIZE=5GB
+
+  TRACING:
```
Let's only enable this part if MINIO_DEBUG is enabled.. this will cause severe readability issues. 

Or bring a '--debug' command line.

hackintoshrao replied:
I'm looking at open tracing and other better tracing tools. Will work on better tracing next
we can choose the approach that time may be. Or we can move it right away. Let me know.

Commit:
instrumentation: instrumentation for locks.
- Intrumentation for locks.
- Detailed test covarage.
- API handler for obtaining lock instrumentation .
